### PR TITLE
Throw warning when intersection Set elements have mismatched types

### PR DESCRIPTION
**Pull request by AlpriElse**
### Problem
While the `type()` function helps give developers visibility of the types of jinjava variables, sometimes developers infer the types of objects by printing them in an expression eg. `{{ some variable }}`.

This can cause some confusion when variables are evaluated to Strings.

For example, within HubL, `{{ content.topic_list }}` is evaluated to a list of strings, but are actually a list of objects.

### Solution
Throw a Warning to inform a developer that the two Sets being intersected have elements of different type to improve DX.

### CODE REVIEW

File: src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java [hunk 1]
```diff
@@ -35,9 +38,41 @@ public Object filter(
...
+        )
+      );
+    }
+
+    return new ArrayList<>(Sets.intersection(varSet, argSet));
+  }
+
+  private String getTypeOfSetElements(Set<Object> set) {
+    if (set.isEmpty()) {
+      return "null";
```
Comment (mattcoley) [inline src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java:73]: Should we make this optional? You should be able to intersect a set with an empty set, however it appears the above logic will add a warning for that.

File: src/test/java/com/hubspot/jinjava/lib/filter/IntersectFilterTest.java [hunk 2]
```diff
@@ -31,4 +34,35 @@ public void itReturnsEmptyOnNullParameters() {
...
+
+    String renderedOutput = interpreter.render("{{ [1, 2, 3]|intersect([1, 2, 3]) }}");
+    assertThat(renderedOutput).isEqualTo("[1, 2, 3]");
+
+    List<TemplateError> errors = interpreter.getErrors();
+    assertThat(errors).isEmpty();
+  }
+
+  @Test
+  public void itThrowsWarningOnMismatchTypes() {
```
Comment (mattcoley) [inline src/test/java/com/hubspot/jinjava/lib/filter/IntersectFilterTest.java:50]: Can we add a unit test for the empty set case to make sure it doesn't add a warning.

File: src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java [hunk 3]
```diff
@@ -35,9 +38,41 @@ public Object filter(
...
   ) {
-    return new ArrayList<>(
-      Sets.intersection(objectToSet(var), objectToSet(parseArgs(interpreter, args)))
-    );
+    Object argObj = parseArgs(interpreter, args);
+
+    Set<Object> varSet = objectToSet(var);
+    Set<Object> argSet = objectToSet(argObj);
+
+    boolean areMismatchedElementTypes = !getTypeOfSetElements(varSet)
```
Comment (mattcoley) [inline src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java:46]: Can we do this in the parent `AbstractSetFilter` class?

File: src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java [hunk 4]
```diff
@@ -35,9 +38,41 @@ public Object filter(
...
+      .equals(getTypeOfSetElements(argSet));
+
+    if (areMismatchedElementTypes) {
+      interpreter.addError(
+        new TemplateError(
+          TemplateError.ErrorType.WARNING,
+          TemplateError.ErrorReason.OTHER,
+          TemplateError.ErrorItem.FILTER,
+          String.format(
+            "Mismatched types. `value` elements are of type `%s` and `list` elements are of type `%s`. This may lead to unexpected behavior.",
```
Comment (mattcoley) [inline src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java:56]: I am slightly confused by this wording. Maybe something like 
`Mismatched Types: input set has elements of type '%s' but arg set has elements of type '%s'. Use `|map` filter to convert sets to the same type for filter to work correctly.`

**Review by mattcoley — CHANGES_REQUESTED** [review]

File: src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java (hunk 1 above)
Comment (AlpriElse) [reply src/main/java/com/hubspot/jinjava/lib/filter/IntersectFilter.java:73]: Kept this existing logic, but reworked the conditional check

**Comment by AlpriElse** [comment]
Moved the warning attachment to `AbstractSetFilter` and implemented throwing the warning in all the implementing classes.

File: src/main/java/com/hubspot/jinjava/lib/filter/AbstractSetFilter.java [hunk 5]
```diff
@@ -29,4 +32,62 @@ protected Set<Object> objectToSet(Object var) {
...
+  ) {
+    Set<Object> varSet = objectToSet(var);
+    Set<Object> argSet = objectToSet(parseArgs(interpreter, args));
+
+    attachMismatchedTypesWarning(interpreter, varSet, argSet);
+
+    return filter(varSet, argSet);
+  }
+
+  public abstract Object filter(Set<Object> varSet, Set<Object> argSet);
```
Comment (AlpriElse) [inline src/main/java/com/hubspot/jinjava/lib/filter/AbstractSetFilter.java:51]: Now that the concrete implementations of `AbstractSetFilter` have the `filter` method is wrapped here, all of them will have the mismatched type warning.

File: src/main/java/com/hubspot/jinjava/lib/filter/AbstractSetFilter.java [hunk 6]
```diff
@@ -29,4 +32,62 @@ protected Set<Object> objectToSet(Object var) {
...
+        interpreter.getLineNumber(),
+        -1,
+        null
+      )
+    );
+  }
+
+  private String getTypeOfSetElements(Set<Object> set) {
+    if (set.isEmpty()) {
+      return "null";
```
[the remark this answers is not in the document]
Comment (AlpriElse) [reply src/main/java/com/hubspot/jinjava/lib/filter/AbstractSetFilter.java:89]: This is borrowed from logic in `TypeFunction` which returns `"null"` as a type when the passed object is null:
https://github.com/HubSpot/jinjava/blob/5cb246370d8f2bdb304fdcfa22cfa57dfab2c14f/src/main/java/com/hubspot/jinjava/lib/fn/TypeFunction.java#L42-L45

File: src/main/java/com/hubspot/jinjava/lib/filter/AbstractSetFilter.java [hunk 7]
```diff
@@ -29,4 +32,59 @@ protected Set<Object> objectToSet(Object var) {
...
+        TemplateError.ErrorReason.OTHER,
+        TemplateError.ErrorItem.FILTER,
+        String.format(
+          "Mismatched Types: input set has elements of type '%s' but arg set has elements of type '%s'. Use |map filter to convert sets to the same type for filter to work correctly.",
+          getTypeOfSetElements(varSet),
+          getTypeOfSetElements(argSet)
+        ),
+        "list",
+        interpreter.getLineNumber(),
+        -1,
```
Comment (mattcoley) [inline src/main/java/com/hubspot/jinjava/lib/filter/AbstractSetFilter.java:81]: You might as well pass in interpreter.getPosition()

**Review by mattcoley — APPROVED** [review]
One more comment and then LGTM
